#!/usr/bin/python3

import htcondor
import os
import sys

from pytest_old.CondorCluster import CondorCluster
from pytest_old.CondorTest import CondorTest
from pytest_old.Globals import *
from pytest_old.HTTPServer import HTTPServer
from pytest_old.PersonalCondor import PersonalCondor
from pytest_old.Utils import Utils

job_cluster_id = None

def SuccessCallback():
    Utils.TLog("Job succeeded unexpectedly when it was supposed to go on hold. Test failed.")
    sys.exit(TEST_FAILURE)

def FailureCallback():
    Utils.TLog("Job went on hold as expected. Test passed")
    sys.exit(TEST_SUCCESS)

def HttpSuccessUrlHandler(request):
    request.protocol_version = "HTTP/1.1"
    request.send_response(200)
    request.send_header("Content-type", "text/html")
    request.end_headers()
    request.wfile.write("<html>Great success!</html>".encode())

def HttpServerErrorUrlHandler(request):
    request.protocol_version = "HTTP/1.1"
    request.send_response(404)


def main():
    test_params = {
        "ENABLE_MULTIFILE_TRANSFER_PLUGINS":    "true",
        "FILETRANSFER_PLUGINS":                 "$(LIBEXEC)/curl_plugin",
        "STARTD_ENVIRONMENT":                  "\"CONDOR_CURL_MAX_RETRY_ATTEMPTS=2\"",
        "STARTER_LOG_NAME_APPEND":              "false"
    }

    # Start a HTTP server
    httpd = HTTPServer()
    httpd.Start()

    # Register three separate URLs
    # Two are supposed to successfully serve a small file
    # One is supposed to generate a 500 server error and cause the job to go on hold
    httpd.RegisterUrlHandler("/file1", HttpSuccessUrlHandler)
    httpd.RegisterUrlHandler("/file2", HttpServerErrorUrlHandler)
    httpd.RegisterUrlHandler("/file3", HttpSuccessUrlHandler)
    file1_url = "dav://127.0.0.1:" + str(httpd._port) + "/file1"
    file2_url = "dav://127.0.0.1:" + str(httpd._port) + "/file2"
    file3_url = "dav://127.0.0.1:" + str(httpd._port) + "/file3"

    # Stand up a PersonalCondor
    personal_handle = CondorTest.StartPersonalCondor("cmd_curl_plugin_multifile_failure", test_params)
    if not personal_handle:
        Utils.TLog("Failed to start the PersonalCondor environment. Exiting.")
        sys.exit(1)
    else:
        Utils.TLog("PersonalCondor environment started successfully")

    # Setup a new test job
    job_args = {
        "executable":               "x_echostring.pl",
        "arguments":                "Testing multifile_curl_plugin failure",
        "output":                   "cmd_curl_plugin_multifile_failure.out",
        "transfer_input_files":     file1_url + ", " + file2_url + ", " + file3_url,
        "should_transfer_files":    "YES"
    }
    test_job = CondorCluster(job_args)

    # Register success and failure handlers.
    test_job.RegisterJobTerminated(SuccessCallback)
    test_job.RegisterJobHeld(FailureCallback)

    # Submit the test job
    result = test_job.Submit()
    job_cluster_id = test_job._cluster_id
    test_job.WaitUntilJobHeld()

    Utils.TLog("Test continued unexpectedly while waiting for job to go on hold. Test failed.")
    sys.exit(TEST_FAILURE)

if __name__ == "__main__":
    # The curl plug-in will respect HTTP_PROXY if it's set, but doing so
    # causes this test to fail when run in a Docker container.  If we
    # want to test this working via a proxy, we should do so explicitly.
    # The second line in this looks super-goofy, but pop() barfs on
    # non-strings and on the empty string, so...
    lowered = dict()
    for k in os.environ:
        lowered[k.lower()] = k
    os.environ.pop(lowered.get("http_proxy", "http_proxy"), None)
    main()
